### PR TITLE
[Kimi K3] Add fused Triton RMSNorm gate

### PR DESCRIPTION
**Pull request by acisseJZhong**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #4832

## Summary

KimiRMSNormGated currently launches separate RMSNorm, weight-scaling, sigmoid, and multiply operations. This adds an opt-in Triton implementation that fuses the forward path and provides custom backward kernels for input, gate, and weight gradients. It preserves the state dict, sharding contract, CPU fallback, and torch.compile support.

This follows the Qwen3.5 OffsetRMSNorm override pattern from https://github.com/pytorch/torchtitan/pull/4417.

Enable with:

```bash
--override.imports torchtitan.overrides.kimi_rmsnorm_gated.triton_kimi_rmsnorm_gated
```

## Testing

- CPU override tests: 4 passed
- GPU numerics, opcheck, and compile tests: 6 passed, 6 subtests passed
- pre-commit: all hooks except Pyrefly passed; Pyrefly was intentionally not run
- Kimi K3 debug pretraining: 8x H100, BF16, FSDP8, sequence length 2048, 16,384 tokens/step, 34 steps

## Preliminary performance

Two repeated steady-state pretraining comparisons (steps 7-34) measured +2.19% and -2.89% throughput, so the debug-model end-to-end result is inconclusive. Peak reserved memory was unchanged at 5.21 GiB.

For a representative large K3 norm shape `[4096, 96, 128]`, the Triton kernel was 4.80x faster in forward (1,257.4 us -> 262.2 us) and 3.17x faster in forward+backward (3,094.9 us -> 975.3 us).

The H100 pretraining experiment used the latest Attention Gym Hopper KDA support and a benchmark-only removal of TorchTitan's existing SM100 capability guard; that unrelated guard change is not included here.

### CODE REVIEW

File: torchtitan/overrides/kimi_gated_rmsnorm.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (acisseJZhong) [inline torchtitan/overrides/kimi_gated_rmsnorm.py:1]: will delete this, just for benchmarking purpose

File: torchtitan/models/common/nn_modules.py [hunk 2]
```diff
@@ -148,6 +151,40 @@ def __init__(self, config: Config):
         )
 
 
+class GatedRMSNorm(Module):
```
Comment (tianyu-l) [inline torchtitan/models/common/nn_modules.py:154]: this file is supposed to host thin wrappers over nn.Modules. This one probably should go to common/norm.py

File: torchtitan/models/common/nn_modules.py [hunk 3]
```diff
@@ -148,6 +151,40 @@ def __init__(self, config: Config):
...
+    """Apply RMSNorm followed by an elementwise unary gate activation."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(Module.Config):
+        dim: int
+        eps: float
+        activation_fn: Callable[[torch.Tensor], torch.Tensor]
```
Comment (tianyu-l) [inline torchtitan/models/common/nn_modules.py:161]: should use what's available in https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/activation.py

File: torchtitan/models/common/nn_modules.py [hunk 4]
```diff
@@ -148,6 +151,40 @@ def __init__(self, config: Config):
...
+        round_normalized_to_input_dtype: bool = False
```
Comment (tianyu-l) [inline torchtitan/models/common/nn_modules.py]: how to decide this?

File: torchtitan/models/kimi_k3/sharding.py [hunk 5]
```diff
@@ -164,15 +164,26 @@ def _set_kda_sharding(
...
+    norm_inputs = {
+        "x_THV": head_placement,
+        "gate_THV": head_placement,
+    }
+    kda_cfg.output_norm.sharding_config = ShardingConfig(
+        state_shardings={"weight": weight_placement},
+        in_src_shardings=norm_inputs,
+        in_dst_shardings=norm_inputs,
+        out_src_shardings=head_placement,
+        out_dst_shardings=head_placement,
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/sharding.py:182]: are these changing input/output placement compared with previously? If so why

File: torchtitan/models/kimi_k3/kda.py [hunk 6]
```diff
@@ -30,28 +29,25 @@
 # W = convolution kernel width.
 
 
-class KimiRMSNormGated(Module):
-    """Per-head RMSNorm followed by a sigmoid output gate."""
+class KimiGatedRMSNorm(GatedRMSNorm):
```
Comment (tianyu-l) [inline torchtitan/models/kimi_k3/kda.py:32]: why creating this module at all. You could've just specify the activation function to be https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/activation.py#L51 in model_registry

File: torchtitan/models/qwen3_5/gdn.py [hunk 7]
```diff
@@ -53,31 +53,26 @@ def _causal_conv1d_varlen(
     return out_BTD.squeeze(0)
 
 
-class RMSNormGated(Module):
-    """Gated RMSNorm: ``silu(gate) * weight * norm(x)``.
-
-    Takes ``(x, gate)`` separately. Weight is ones-initialized.
-    """
+class Qwen35GatedRMSNorm(GatedRMSNorm):
```
Comment (tianyu-l) [inline torchtitan/models/qwen3_5/gdn.py:56]: similar

File: torchtitan/overrides/compiled_gated_rmsnorm.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    "compiled_qwen35_gated_rmsnorm",
+]
+
+
+class CompiledGatedRMSNorm(GatedRMSNorm):
+    """RMSNorm with a compiled configurable unary output gate."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(GatedRMSNorm.Config):
+        activation_fn: Callable[[torch.Tensor], torch.Tensor] = torch.sigmoid
```
Comment (tianyu-l) [inline torchtitan/overrides/compiled_gated_rmsnorm.py:38]: not clear why need this

File: torchtitan/overrides/compiled_gated_rmsnorm.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        self,
+        x: torch.Tensor,
+        gate: torch.Tensor,
+    ) -> torch.Tensor:
+        with maybe_regional_inductor(self.inductor_options):
+            return self._compiled_gated_rms_norm(x, gate)
+
+
+@override(
+    target=KimiGatedRMSNorm.Config,
```
Comment (tianyu-l) [inline torchtitan/overrides/compiled_gated_rmsnorm.py:67]: one general override is enough

**Review by tianyu-l — CHANGES_REQUESTED** [review]
nice

File: torchtitan/models/common/nn_modules.py (hunk 4 above)
Comment (acisseJZhong) [reply torchtitan/models/common/nn_modules.py]: sorry, updated a new version.

File: torchtitan/models/common/nn_modules.py (hunk 2 above)
Comment (acisseJZhong) [reply torchtitan/models/common/nn_modules.py:154]: we don't have a common/norm.py yet, let me create one

**Review by felipemello1 — COMMENTED** [review]
In my mind, we shouldn't need a function for every compile override, i.e. we should **not** have `compiled_qwen35_gated_rmsnorm`, `compiled_kimi_gated_rmsnorm`, etc.

Instead, just a generic `compile_decorator_override` and we pass  the name of the modules we want to compile. Is that possible?

File: torchtitan/models/kimi_k3/sharding.py (hunk 5 above)
Comment (acisseJZhong) [reply torchtitan/models/kimi_k3/sharding.py]: not changing values, add this because we need to replace here https://github.com/pytorch/torchtitan/blob/08f29e75fe6d7d95c8a109f3c2791ad59942c6d3/torchtitan/overrides/kimi_gated_rmsnorm.py#L384

**Comment by acisseJZhong** [comment]
> just a generic compile_decorator_override and we pass the name of the modules we want to compile

@felipemello1 yup changed towards this after abstracting out activation config.

File: torchtitan/models/flux/model/layers.py [hunk 10]
```diff
@@ -406,7 +406,7 @@ def __init__(self, config: Config):
         ).build()
         self.linear = config.linear.build()
         self.adaLN_modulation = Sequential(
-            SiLU.Config().build(),
+            nn.SiLU(),
```
Comment (acisseJZhong) [inline torchtitan/models/flux/model/layers.py]: this is due to I removed the configurable SiLu from nn.module

File: torchtitan/distributed/compile.py [hunk 11]
```diff
@@ -107,31 +108,34 @@ def _maybe_enable_async_tp(
...
+    be scooped here and raise rather than silently degrading.
     """
-    from torchtitan.models.common.attention import FlexInnerAttention
-
-    uses_flex = any(isinstance(m, FlexInnerAttention) for m in model.modules())
-    # Non-flex models never need the scoop; the default inductor backend already
-    # lowers the flex region directly. Both are left on the unmodified backend.
-    if not uses_flex or backend == "inductor":
+    has_inductor_region = any(
+        getattr(module, "_has_inductor_region", False) for module in model.modules()
```
Comment (tianyu-l) [inline torchtitan/distributed/compile.py]: Using a private class variable to indicate inductor usage or not -- I'm OK with this for now, but maybe it's worth to invent some e.g. function wrapper based solution, later.

Maybe add a TODO here.

cc @anijain2305

File: torchtitan/overrides/kimi_gated_rmsnorm.py
Comment (tianyu-l) [inline torchtitan/overrides/kimi_gated_rmsnorm.py]: don't forget removing

**Review by tianyu-l — APPROVED** [review]

File: torchtitan/overrides/compiled_gated_rmsnorm.py
Comment (tianyu-l) [inline torchtitan/overrides/compiled_gated_rmsnorm.py]: let's enable this override at least in every RL config (maybe other than the batch invariant ones if the compiler-generated kernels are not)

File: torchtitan/overrides/compiled_gated_rmsnorm.py [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+class CompiledGatedRMSNorm(GatedRMSNorm):
+    """RMSNorm with a compiled configurable unary output gate."""
+
+    @dataclass(kw_only=True, slots=True)
+    class Config(GatedRMSNorm.Config):
+        pass
+
+    _has_inductor_region: ClassVar[bool] = True
+    inductor_options: ClassVar[dict[str, bool]] = {
```
Comment (acisseJZhong) [inline torchtitan/overrides/compiled_gated_rmsnorm.py]: wondering what's the best recommendation settings for max_autotune and coordinate_descent_tuning? Flex attention defaults it to True but the comment seems suggest otherwise? https://github.com/pytorch/torchtitan/blob/46ac2d9354d67b8cfb4cfee296759e677781a832/torchtitan/models/common/attention.py#L249
cc @SherlockNoMad @drisspg if you know

File: torchtitan/overrides/compiled_gated_rmsnorm.py
Comment (acisseJZhong) [reply torchtitan/overrides/compiled_gated_rmsnorm.py]: added to every RL config that uses GatedRMSNorm and non batch invariant. Tested the compiled kernel is deterministic, but is not batch invariant(specifically grad_weight is not). I remembered @drisspg said there's some flag to set to enable batch inv?
